cmd/registry/controller/resource-name.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"github.com/apigee/registry/server/registry/names"
+)
+
+// This interface is used to describe resource patterns
+// Example:
+// projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml
+// projects/demo/locations/global/apis/-/versions/-/specs/-/artifacts/-
+type ResourceName interface {
+	GetArtifact() string
+	GetSpec() string
+	GetVersion() string
+	GetApi() string
+	GetName() string
+}
+
+type SpecName struct {
+	Spec names.Spec
+}
+
+func (s SpecName) GetArtifact() string {
+	return ""
+}
+
+func (s SpecName) GetSpec() string {
+	return s.Spec.String()
+}
+
+func (s SpecName) GetVersion() string {
+	return s.Spec.Version().String()
+}
+
+func (s SpecName) GetApi() string {
+	return s.Spec.Api().String()
+}
+
+func (s SpecName) GetName() string {
+	return s.Spec.String()
+}
+
+type VersionName struct {
+	Version names.Version
+}
+
+func (v VersionName) GetArtifact() string {
+	return ""
+}
+
+func (v VersionName) GetSpec() string {
+	return ""
+}
+
+func (v VersionName) GetVersion() string {
+	return v.Version.String()
+}
+
+func (v VersionName) GetApi() string {
+	return v.Version.Api().String()
+}
+
+func (v VersionName) GetName() string {
+	return v.Version.String()
+}
+
+type ApiName struct {
+	Api names.Api
+}
+
+func (a ApiName) GetArtifact() string {
+	return ""
+}
+
+func (a ApiName) GetSpec() string {
+	return ""
+}
+
+func (a ApiName) GetVersion() string {
+	return ""
+}
+
+func (a ApiName) GetApi() string {
+	return a.Api.String()
+}
+
+func (a ApiName) GetName() string {
+	return a.Api.String()
+}
+
+type ArtifactName struct {
+	Artifact names.Artifact
+}
+
+func (ar ArtifactName) GetArtifact() string {
+	return ar.Artifact.String()
+}
+
+func (ar ArtifactName) GetSpec() string {
+	specName := names.Spec{
+		ProjectID: ar.Artifact.ProjectID(),
+		ApiID:     ar.Artifact.ApiID(),
+		VersionID: ar.Artifact.VersionID(),
+		SpecID:    ar.Artifact.SpecID(),
+	}
+	// if err := specName.Validate(); err != nil {
```

Comment:
@timburks 

I currently can't validate the generated specName because `ValidateID()` functions currently doesn't support `-` hyphens, and thus `Validate()` not supporting such patterns `projects/demo/locations/global/apis/-/versions/-/specs/-`. Am I using this correctly, or am I missing anything here?

Reply:
It might be better to call this a "spec pattern" than a "spec name", since it can match multiple specs. The `ValidateID()` function was written to validate actual ids (not patterns) and I think its current usage expects that.

Instead of calling specName.Validate(), you could validate this pattern by passing it to the [`names.ParseSpec`](https://github.com/apigee/registry/blob/cb70a1d794d589db434c0097af98ae471d9d01f4/server/registry/names/spec.go#L127) function, which (I believe) accepts wildcards.

Reply:
Thanks that makes sense. I can use `ParseSpec` to validate the generated name.

While this issue is not blocking, I realized another thing in the names package behavior:
```
spec, err := names.ParseSpec("projects/demo/locations/global/apis/-/versions/-/specs/-")
	if err != nil {
		fmt.Printf("ERROR PARSE: %s", err)
	}
	err = spec.Validate()
	if err != nil {
		fmt.Printf("ERROR VALIDATE: %s", err)
	}
```

If I execute the above code snippet, `ParseSpec()` allows the presence of `-` hyphens whereas, `Validate()` doesn't support that. This behavior feels a little confusing to me. Either both or non of them should support the use of `-` hyphens.